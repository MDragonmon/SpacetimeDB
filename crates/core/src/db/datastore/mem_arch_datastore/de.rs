//! Provides the function [`read_row_from_page(hasher, page, fixed_offset, ty)`]
//! which reads `value = page.get_row_data(fixed_offset, fixed_row_size)` typed at `ty`
//! and associated var len objects in `value` into a `ProductValue`.

use super::{
    blob_store::BlobStore,
    indexes::{Bytes, PageOffset},
    layout::{
        align_to, AlgebraicTypeLayout, HasLayout as _, ProductTypeLayout, RowTypeLayout, SumTypeLayout, VarLenType,
    },
    page::Page,
    var_len::VarLenRef,
};
use core::mem;
use spacetimedb_sats::{AlgebraicValue, ProductValue};

/// Reads the row in `page` where the fixed part starts at `fixed_offset`
/// and lasts `ty.size()` bytes. This region is typed at `ty`.
///
/// # Safety
///
/// 1. the `fixed_offset` must point at a row in `page` lasting `ty.size()` byte.
/// 2. the row must be a valid `ty`.
/// 3. for any `vlr: VarLenRef` stored in the row,
///   `vlr.first_offset` must either be `NULL` or point to a valid granule in `page`.
pub unsafe fn read_row_from_page(
    page: &Page,
    blob_store: &dyn BlobStore,
    fixed_offset: PageOffset,
    ty: &RowTypeLayout,
) -> ProductValue {
    let fixed_bytes = page.get_row_data(fixed_offset, ty.size());
    // SAFETY:
    // - Per 1. and 2., `fixed_bytes` points at a row in `page` valid for `ty`.
    // - Per 3., for any `vlr: VarLenRef` stored in `fixed_bytes`,
    //   `vlr.first_offset` is either `NULL` or points to a valid granule in `page`.
    unsafe { deserialize_product(fixed_bytes, page, blob_store, &mut 0, ty.product()) }
}

/// Reads every product field in `value = &bytes[range_move(0..ty.size(), *curr_offset)]`,
/// which is typed at `ty`,
/// and then stitches together.
///
/// SAFETY:
/// 1. the `value` must be valid at type `ty` and properly aligned for `ty`.
/// 2. for any `vlr: VarLenRef` stored in `value`,
///   `vlr.first_offset` must either be `NULL` or point to a valid granule in `page`.
unsafe fn deserialize_product(
    bytes: &Bytes,
    page: &Page,
    blob_store: &dyn BlobStore,
    curr_offset: &mut usize,
    ty: &ProductTypeLayout,
) -> ProductValue {
    let elements = ty
        .elements
        .iter()
        // SAFETY: By 1., `value` is valid at `ty`,
        // so it follows that valid and properly aligned sub-`value`s
        // are valid `elem_ty.ty`s.
        // By 2., and the above, it follows that sub-`value`s won't have dangling `VarLenRef`s.
        .map(|elem_ty| unsafe { deserialize_value(bytes, page, blob_store, curr_offset, &elem_ty.ty) })
        .collect::<Vec<AlgebraicValue>>();
    ProductValue { elements }
}

/// Reads the tag of the sum value and selects the data variant type.
///
/// # Safety
///
/// `bytes[curr_offset..]` has a sum value typed at `ty`.
pub unsafe fn read_tag<'ty>(
    bytes: &Bytes,
    ty: &'ty SumTypeLayout,
    curr_offset: usize,
) -> (u8, &'ty AlgebraicTypeLayout) {
    let tag_offset = ty.offset_of_tag();
    let tag = bytes[curr_offset + tag_offset];
    // SAFETY: Caller promised that `bytes[curr_offset..]` has a sum value typed at `ty`.
    // We can therefore assume that `curr_offset + tag_offset` refers to a valid `u8`.
    let tag = unsafe { tag.assume_init() };

    // Extract the variant data type depending on the tag.
    let data_ty = &ty.variants[tag as usize].ty;

    (tag, data_ty)
}

/// Reads `value = &bytes[range_move(0..ty.size(), *curr_offset)]`
/// into an `AlgebraicValue` typed at `ty`
/// using `blob_store` to retreive the bytes of any large blob object
/// and `curr_offset`, then advanced, to decide where to start reading.
///
/// SAFETY:
/// 1. the `value` must be valid at type `ty` and properly aligned for `ty`.
/// 2. for any `vlr: VarLenRef` stored in `value`,
///   `vlr.first_offset` must either be `NULL` or point to a valid granule in `page`.
unsafe fn deserialize_value(
    bytes: &Bytes,
    page: &Page,
    blob_store: &dyn BlobStore,
    curr_offset: &mut usize,
    ty: &AlgebraicTypeLayout,
) -> AlgebraicValue {
    let ty_alignment = ty.align();
    *curr_offset = align_to(*curr_offset, ty_alignment);

    let res = match ty {
        AlgebraicTypeLayout::Sum(ty) => {
            // Read the tag of the sum value.
            // SAFETY: `bytes[curr_offset..]` hold a sum value at `ty`.
            let (tag, data_ty) = unsafe { read_tag(bytes, ty, *curr_offset) };

            // Read the variant data value.
            let mut data_offset = *curr_offset + ty.offset_of_variant_data(tag);
            // SAFETY: `value` is valid at `ty` so given `tag`,
            // we know `data_value = &bytes[range_move(0..data_ty.size(), data_offset))`
            // is valid at `data_ty`.
            // By 2., and the above, we also know that `data_value` won't have dangling `VarLenRef`s.
            let data_av = unsafe { deserialize_value(bytes, page, blob_store, &mut data_offset, data_ty) };
            *curr_offset += ty.size();

            // Stitch together.
            AlgebraicValue::sum(tag, data_av)
        }
        AlgebraicTypeLayout::Product(ty) => {
            // SAFETY: `value` was valid at `ty` and `VarLenRef`s won't be dangling.
            unsafe { deserialize_product(bytes, page, blob_store, curr_offset, ty) }.into()
        }
        // The primitive types:
        //
        // SAFETY (applies to app primitive types):
        // Per caller requirement, know `value` points to a valid `ty`.
        // Thus `&bytes[range_move(0..ty.size(), *curr_offset)]` points to init bytes
        // and `ty.size()` corresponds exactly to `N = 1, 1, 1, 2, 2, 4, 4, 8, 8, 16, 16, 4, 8`.
        &AlgebraicTypeLayout::Bool => (unsafe { read_u8_unchecked(bytes, curr_offset) } != 0).into(),
        &AlgebraicTypeLayout::I8 => (unsafe { read_u8_unchecked(bytes, curr_offset) } as i8).into(),
        &AlgebraicTypeLayout::U8 => unsafe { read_u8_unchecked(bytes, curr_offset) }.into(),
        &AlgebraicTypeLayout::I16 => i16::from_le_bytes(unsafe { read_byte_array(bytes, curr_offset) }).into(),
        &AlgebraicTypeLayout::U16 => u16::from_le_bytes(unsafe { read_byte_array(bytes, curr_offset) }).into(),
        &AlgebraicTypeLayout::I32 => i32::from_le_bytes(unsafe { read_byte_array(bytes, curr_offset) }).into(),
        &AlgebraicTypeLayout::U32 => u32::from_le_bytes(unsafe { read_byte_array(bytes, curr_offset) }).into(),
        &AlgebraicTypeLayout::I64 => i64::from_le_bytes(unsafe { read_byte_array(bytes, curr_offset) }).into(),
        &AlgebraicTypeLayout::U64 => u64::from_le_bytes(unsafe { read_byte_array(bytes, curr_offset) }).into(),
        &AlgebraicTypeLayout::I128 => i128::from_le_bytes(unsafe { read_byte_array(bytes, curr_offset) }).into(),
        &AlgebraicTypeLayout::U128 => u128::from_le_bytes(unsafe { read_byte_array(bytes, curr_offset) }).into(),
        &AlgebraicTypeLayout::F32 => f32::from_le_bytes(unsafe { read_byte_array(bytes, curr_offset) }).into(),
        &AlgebraicTypeLayout::F64 => f64::from_le_bytes(unsafe { read_byte_array(bytes, curr_offset) }).into(),

        // The var-len cases.
        &AlgebraicTypeLayout::String => {
            // SAFETY: `value` was valid at and aligned for `::String`
            // which stores a `vlr: VarLenRef` as its fixed value.
            // The range thus is valid and properly aligned for `VarLenRef`.
            // Moreover, `vlr.first_granule` was promised by the caller
            // to either be `NULL` or point to a valid granule in `page`.
            let bytes = unsafe { read_vlr_to_bytes(page, blob_store, bytes, curr_offset) };
            let string = String::from_utf8(bytes).unwrap();
            string.into()
        }
        AlgebraicTypeLayout::VarLen(VarLenType::Array(ty) | VarLenType::Map(ty)) => {
            // SAFETY: `value` was valid at and aligned for `ty`.
            // These `ty` store a `vlr: VarLenRef` as their fixed value.
            // The range thus is valid and properly aligned for `VarLenRef`.
            // Moreover, `vlr.first_granule` was promised by the caller
            // to either be `NULL` or point to a valid granule in `page`.
            let bytes = unsafe { read_vlr_to_bytes(page, blob_store, bytes, curr_offset) };
            AlgebraicValue::decode(ty, &mut &*bytes).unwrap()
        }
    };
    // TODO(perf,bikeshedding): unncessary work for some cases?
    *curr_offset = align_to(*curr_offset, ty_alignment);
    res
}

/// Read the bytes of a var-len object
/// referred to at by the var-len reference at the current offset.
///
/// SAFETY: (1) `data = &bytes[range_move(0..size_of::<VarLenRef>(), *curr_offset)]`
/// must point to a valid `VarLenRef` and `data` must be properly aligned for a `VarLenRef`.
/// (2) The `vlr.first_granule` must be `NULL` or must point to a valid granule in `page`.
unsafe fn read_vlr_to_bytes(
    page: &Page,
    blob_store: &dyn BlobStore,
    bytes: &Bytes,
    curr_offset: &mut usize,
) -> Vec<u8> {
    // SAFETY: (1) We have a valid `VarLenRef` at `&data`.
    let vlr = unsafe { read_vlr(bytes, curr_offset) };
    if vlr.is_large_blob() {
        // SAFETY: (2)
        unsafe { read_vec_from_blob(page, blob_store, vlr) }
    } else {
        // SAFETY: (2)
        unsafe { read_vec(page, vlr) }
    }
}

/// Read the bytes of the var-len object referred to by `vlr`.
///
/// SAFETY: (1) `vlr.first_granule` must be `NULL` or must point to a valid granule in `page`.
unsafe fn read_vec(page: &Page, vlr: VarLenRef) -> Vec<u8> {
    let mut bytes = Vec::with_capacity(vlr.length_in_bytes as usize);
    // An initialized var-len object is either a UTF-8 string or BSATN-encoded,
    // and in either case is fully init.
    //
    // SAFETY: (1)
    let var_iter = unsafe { page.iter_vlo_data(vlr.first_granule) };
    for granule in var_iter {
        bytes.extend(granule);
    }
    bytes
}

/// Read a large blob object from `vlr.first_granule`.
///
/// SAFETY: (1) `vlr.first_granule` must be `NULL` or must point to a valid granule in `page`.
#[cold]
#[inline(never)]
unsafe fn read_vec_from_blob(page: &Page, blob_store: &dyn BlobStore, vlr: VarLenRef) -> Vec<u8> {
    // SAFETY: (1)
    let mut var_iter = unsafe { page.iter_var_len_object(vlr.first_granule) };
    let granule = var_iter.next().unwrap();
    let hash = granule.blob_hash();
    blob_store.retrieve_blob(&hash).unwrap().to_owned()
}

/// Read a [`VarLenRef`] from `bytes` at the `curr_offset` and advance.
///
/// # Safety
///
/// `data = &bytes[range_move(0..size_of::<VarLenRef>(), *curr_offset)]`
/// must point to a valid `VarLenRef` and `data` must be properly aligned for a `VarLenRef`.
pub unsafe fn read_vlr(bytes: &Bytes, curr_offset: &mut usize) -> VarLenRef {
    // SAFETY: `T = VarLenRef` and caller promised that `data` points to a valid `VarLenRef`.
    unsafe { read_from_bytes(bytes, curr_offset) }
}

/// Read a byte from `bytes` at the `curr_offset` and advance.
///
/// # Safety
///
/// Let `value = &bytes[range_move(0..1, *curr_offset)]`.
/// Then `value` must point to a valid `u8`.
/// Proper alignment is already trivially satisfied as `align_of::<u8>() == 1`.
pub unsafe fn read_u8_unchecked(bytes: &Bytes, curr_offset: &mut usize) -> u8 {
    unsafe { read_byte_array::<1>(bytes, curr_offset)[0] }
}

/// Read a byte array of `N` elements from `bytes` at the `curr_offset` and advance.
///
/// # Safety
///
/// Let `value = &bytes[range_move(0..N, *curr_offset)]`.
/// Then `value` must point to a valid `[u8; N]`.
/// Proper alignment is already trivially satisfied as `align_of::<u8>() == 1`.
pub unsafe fn read_byte_array<const N: usize>(bytes: &Bytes, curr_offset: &mut usize) -> [u8; N] {
    unsafe { read_from_bytes(bytes, curr_offset) }
}

/// Read a `T` from `bytes` at the `curr_offset` and advance by `size` bytes.
///
/// SAFETY: Let `value = &bytes[range_move(0..size_of::<T>(), *curr_offset)]`.
/// Then `value` must point to a valid `T` and must be properly aligned for `T`.
unsafe fn read_from_bytes<T: Copy>(bytes: &Bytes, curr_offset: &mut usize) -> T {
    let bytes = &bytes[*curr_offset..];
    *curr_offset += mem::size_of::<T>();
    let ptr: *const T = bytes.as_ptr().cast();
    // SAFETY: Caller promised that `ptr` points to a `T`.
    // Moreover, `ptr` is derived from a shared reference with permission to read this range.
    unsafe { *ptr }
}
