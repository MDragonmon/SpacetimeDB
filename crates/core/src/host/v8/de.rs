#![allow(dead_code)]

use super::error::{exception_already_thrown, ExceptionThrown};
use super::from_value::{cast, FromValue};
use core::fmt;
use core::iter::{repeat_n, RepeatN};
use core::mem::MaybeUninit;
use derive_more::From;
use spacetimedb_sats::de::{
    self, ArrayVisitor, DeserializeSeed, NoneAccess, ProductVisitor, SliceVisitor, SomeAccess, SumVisitor,
};
use spacetimedb_sats::{i256, u256};
use std::borrow::{Borrow, Cow};
use v8::{Array, Global, HandleScope, Local, Name, Object, Uint8Array, Value};

/// Deserializes from V8 values.
pub(super) struct Deserializer<'a, 's> {
    common: DeserializerCommon<'a, 's>,
    input: Local<'s, Value>,
}

impl<'a, 's> Deserializer<'a, 's> {
    /// Creates a new deserializer from `input` in `scope`.
    pub fn new(scope: &'a mut HandleScope<'s>, input: Local<'_, Value>, key_cache: &'a mut KeyCache) -> Self {
        let input = Local::new(scope, input);
        let common = DeserializerCommon { scope, key_cache };
        Deserializer { input, common }
    }
}

/// Things shared between various [`Deserializer`]s.
///
/// The lifetime `'s` is that of the scope of values deserialized.
struct DeserializerCommon<'a, 's> {
    /// The scope of values to deserialize.
    scope: &'a mut HandleScope<'s>,
    /// A cache for frequently used strings.
    key_cache: &'a mut KeyCache,
}

impl<'s> DeserializerCommon<'_, 's> {
    fn reborrow(&mut self) -> DeserializerCommon<'_, 's> {
        DeserializerCommon {
            scope: self.scope,
            key_cache: self.key_cache,
        }
    }
}

/// The possible errors that [`Deserializer`] can produce.
#[derive(Debug, From)]
pub(super) enum Error<'s> {
    Value(Local<'s, Value>),
    Exception(ExceptionThrown),
    Custom(String),
}

impl de::Error for Error<'_> {
    fn custom(msg: impl fmt::Display) -> Self {
        Self::Custom(msg.to_string())
    }
}

/// Returns a scratch buffer to fill when deserializing strings.
fn scratch_buf<const N: usize>() -> [MaybeUninit<u8>; N] {
    [const { MaybeUninit::uninit() }; N]
}

/// A cache for frequently used strings to avoid re-interning them.
#[derive(Default)]
pub(super) struct KeyCache {
    /// The `tag` property for sum values in JS.
    tag: Option<Global<v8::String>>,
    /// The `value` property for sum values in JS.
    value: Option<Global<v8::String>>,
}

impl KeyCache {
    /// Returns the `tag` property name.
    pub(super) fn tag<'s>(&mut self, scope: &mut HandleScope<'s>) -> Local<'s, v8::String> {
        Self::get_or_create_key(scope, &mut self.tag, "tag")
    }

    /// Returns the `value` property name.
    pub(super) fn value<'s>(&mut self, scope: &mut HandleScope<'s>) -> Local<'s, v8::String> {
        Self::get_or_create_key(scope, &mut self.value, "value")
    }

    /// Returns an interned string corresponding to `string`
    /// and memoizes the creation on the v8 side.
    fn get_or_create_key<'s>(
        scope: &mut HandleScope<'s>,
        slot: &mut Option<Global<v8::String>>,
        string: &str,
    ) -> Local<'s, v8::String> {
        if let Some(s) = &*slot {
            v8::Local::new(scope, s)
        } else {
            let s = v8_interned_string(scope, string);
            *slot = Some(v8::Global::new(scope, s));
            s
        }
    }
}

// Creates an interned [`v8::String`].
pub(super) fn v8_interned_string<'s>(scope: &mut HandleScope<'s>, field: &str) -> Local<'s, v8::String> {
    // Internalized v8 strings are significantly faster than "normal" v8 strings
    // since v8 deduplicates re-used strings minimizing new allocations
    // see: https://github.com/v8/v8/blob/14ac92e02cc3db38131a57e75e2392529f405f2f/include/v8.h#L3165-L3171
    v8::String::new_from_utf8(scope, field.as_ref(), v8::NewStringType::Internalized).unwrap()
}

/// Extracts a reference `&'s T` from an owned V8 [`Local<'s, T>`].
///
/// The lifetime `'s` is that of the [`HandleScope<'s>`].
/// This ensures that the reference to `T` won't outlive the `HandleScope`.
fn deref_local<'s, T>(local: Local<'s, T>) -> &'s T {
    let reference = local.borrow();
    // SAFETY: Lifetime extend `'0` to `'s`.
    // This is safe as the returned reference `&'s T`
    // will not outlive its `HandleScope<'s, _>`,
    // as both are tied to the lifetime `'s`.
    unsafe { core::mem::transmute::<&T, &'s T>(reference) }
}

/// Deserializes a primitive via [`FromValue`].
macro_rules! deserialize_primitive {
    ($dmethod:ident, $t:ty) => {
        fn $dmethod(self) -> Result<$t, Self::Error> {
            FromValue::from_value(self.input, self.common.scope).map_err(Error::Value)
        }
    };
}

impl<'de, 'a, 's: 'de> de::Deserializer<'de> for Deserializer<'a, 's> {
    type Error = Error<'s>;

    // Deserialization of primitive types defers to `FromValue`.
    deserialize_primitive!(deserialize_bool, bool);
    deserialize_primitive!(deserialize_u8, u8);
    deserialize_primitive!(deserialize_u16, u16);
    deserialize_primitive!(deserialize_u32, u32);
    deserialize_primitive!(deserialize_u64, u64);
    deserialize_primitive!(deserialize_u128, u128);
    deserialize_primitive!(deserialize_u256, u256);
    deserialize_primitive!(deserialize_i8, i8);
    deserialize_primitive!(deserialize_i16, i16);
    deserialize_primitive!(deserialize_i32, i32);
    deserialize_primitive!(deserialize_i64, i64);
    deserialize_primitive!(deserialize_i128, i128);
    deserialize_primitive!(deserialize_i256, i256);
    deserialize_primitive!(deserialize_f64, f64);
    deserialize_primitive!(deserialize_f32, f32);

    fn deserialize_product<V: ProductVisitor<'de>>(self, visitor: V) -> Result<V::Output, Self::Error> {
        let object = cast!(
            self.common.scope,
            self.input,
            Object,
            "object for product type `{}`",
            visitor.product_name().unwrap_or("<unknown>")
        )?;

        visitor.visit_named_product(ProductAccess {
            common: self.common,
            object,
            next_value: None,
            index: 0,
        })
    }

    fn deserialize_sum<V: SumVisitor<'de>>(self, visitor: V) -> Result<V::Output, Self::Error> {
        let scope = &mut *self.common.scope;
        let sum_name = visitor.sum_name().unwrap_or("<unknown>");

        // We expect a canonical representation of a sum value in JS to be
        // `{ tag: "foo", value: a_value_for_foo }`
        // with special convenience for optionals
        // where we also accept `null`, `undefined` and an object without `tag`.
        let (object, tag_field) = 'treat_as_regular_sum: {
            // Optionals receive some special handling for added convenience in JS.
            if visitor.is_option() {
                // If we don't have an object at all,
                // it's either `null | undefined` which means `none`
                // or it is `some(the_value)`.
                if let Some(object) = self.input.to_object(scope) {
                    // If there is `tag` field, treat this as a normal sum.
                    // Otherwise, we have `some(the_value)`.
                    let tag_field = self.common.key_cache.tag(scope);
                    if object
                        .has_own_property(scope, tag_field.into())
                        .ok_or_else(exception_already_thrown)?
                    {
                        break 'treat_as_regular_sum (object, tag_field);
                    }
                } else if self.input.is_null_or_undefined() {
                    // JS has support for `undefined` and `null` values.
                    // It's reasonable to interpret these as `None`
                    // when we're deserializing to an optional value
                    // rust-side, such as `Option<T>`.
                    return visitor.visit_sum(NoneAccess::new());
                }

                return visitor.visit_sum(SomeAccess::new(self));
            } else {
                let tag_field = self.common.key_cache.tag(scope);
                let val = cast!(scope, self.input, Object, "object for sum type `{}`", sum_name)?;
                (val, tag_field)
            }
        };

        // Extract the `tag` field. It needs to contain a string.
        let tag = object
            .get(scope, tag_field.into())
            .ok_or_else(exception_already_thrown)?;
        let tag = cast!(scope, tag, v8::String, "string for sum tag of `{}`", sum_name)?;

        // Extract the `value` field.
        let value_field = self.common.key_cache.value(scope);
        let value = object
            .get(scope, value_field.into())
            .ok_or_else(exception_already_thrown)?;

        // Stitch it all together.
        visitor.visit_sum(SumAccess {
            common: self.common,
            tag,
            value,
        })
    }

    fn deserialize_str<V: SliceVisitor<'de, str>>(self, visitor: V) -> Result<V::Output, Self::Error> {
        let val = cast!(self.common.scope, self.input, v8::String, "`string`")?;
        let mut buf = scratch_buf::<64>();
        match val.to_rust_cow_lossy(self.common.scope, &mut buf) {
            Cow::Borrowed(s) => visitor.visit(s),
            Cow::Owned(string) => visitor.visit_owned(string),
        }
    }

    fn deserialize_bytes<V: SliceVisitor<'de, [u8]>>(self, visitor: V) -> Result<V::Output, Self::Error> {
        let arr = cast!(self.common.scope, self.input, Uint8Array, "`Uint8Array` for bytes")?;
        let storage: &'static mut [u8] = &mut [0; v8::TYPED_ARRAY_MAX_SIZE_IN_HEAP];
        let bytes: &'s [u8] = deref_local(arr).get_contents(storage);
        visitor.visit_borrowed(bytes)
    }

    fn deserialize_array_seed<V: ArrayVisitor<'de, T::Output>, T: DeserializeSeed<'de> + Clone>(
        self,
        visitor: V,
        seed: T,
    ) -> Result<V::Output, Self::Error> {
        let arr = cast!(self.common.scope, self.input, v8::Array, "`Array`")?;
        visitor.visit(ArrayAccess::new(arr, self.common, seed))
    }
}

/// Provides access to the field names and values in a JS object
/// under the assumption that it's a product.
struct ProductAccess<'a, 's> {
    common: DeserializerCommon<'a, 's>,
    /// The input object being deserialized.
    object: Local<'s, Object>,
    /// A field's value, to deserialize next in [`NamedProductAccess::get_field_value_seed`].
    next_value: Option<Local<'s, Value>>,
    /// The index in the product to
    index: usize,
}

/// Normalizes `field` into an interned `v8::String`.
pub(super) fn intern_field_name<'s>(scope: &mut HandleScope<'s>, field: Option<&str>, index: usize) -> Local<'s, Name> {
    let field = match field {
        Some(field) => Cow::Borrowed(field),
        None => Cow::Owned(format!("{index}")),
    };
    v8_interned_string(scope, field).into()
}

impl<'de, 's: 'de> de::NamedProductAccess<'de> for ProductAccess<'_, 's> {
    type Error = Error<'s>;

    fn get_field_ident<V: de::FieldNameVisitor<'de>>(&mut self, visitor: V) -> Result<Option<V::Output>, Self::Error> {
        let scope = &mut *self.common.scope;
        let mut field_names = visitor.field_names();
        while let Some(field) = field_names.nth(self.index) {
            // Get and advance the current index.
            let index = self.index;
            self.index += 1;

            // Normalize the field name.
            // Integer keys are converted to strings,
            // as that is supported on JS objects.
            let key = intern_field_name(scope, field, index);

            // Check that such a field/key exists.
            if !self
                .object
                .has_own_property(scope, key)
                .ok_or_else(exception_already_thrown)?
            {
                continue;
            }

            // Extract the next value, to be processed in `get_field_value_seed`.
            let val = self
                .object
                .get(scope, key.into())
                .ok_or_else(exception_already_thrown)?;
            self.next_value = Some(val);

            drop(field_names);
            return Ok(Some(visitor.visit_seq(index)));
        }

        Ok(None)
    }

    fn get_field_value_seed<T: de::DeserializeSeed<'de>>(&mut self, seed: T) -> Result<T::Output, Self::Error> {
        let common = self.common.reborrow();
        // Extract the field's value.
        let input = self
            .next_value
            .take()
            .expect("Call next_key_seed before next_value_seed");
        // Deserialize the field's value.
        seed.deserialize(Deserializer { common, input })
    }
}

/// Used in `Deserializer::deserialize_sum` to translate a `tag` property of a JS object
/// to a variant and to provide a deserializer for its value/payload.
struct SumAccess<'a, 's> {
    common: DeserializerCommon<'a, 's>,
    /// The tag of the sum value.
    tag: Local<'s, v8::String>,
    /// The value of the sum value.
    value: Local<'s, Value>,
}

impl<'de, 'a, 's: 'de> de::SumAccess<'de> for SumAccess<'a, 's> {
    type Error = Error<'s>;
    type Variant = Deserializer<'a, 's>;

    fn variant<V: de::VariantVisitor<'de>>(self, visitor: V) -> Result<(V::Output, Self::Variant), Self::Error> {
        // Read the `tag` property in JS.
        // We generously provide 32 bytes for inline reading of the property
        // before resorting to the heap.
        let mut buf = scratch_buf::<32>();
        let name = self.tag.to_rust_cow_lossy(self.common.scope, &mut buf);

        // Select the variant to deserialize.
        let variant = visitor.visit_name::<Self::Error>(&name)?;

        // Prepare the deserialization of the payload.
        let dpayload = Deserializer {
            common: self.common,
            input: self.value,
        };

        Ok((variant, dpayload))
    }
}

impl<'de, 'a, 's: 'de> de::VariantAccess<'de> for Deserializer<'a, 's> {
    type Error = Error<'s>;

    fn deserialize_seed<T: de::DeserializeSeed<'de>>(self, seed: T) -> Result<T::Output, Self::Error> {
        seed.deserialize(self)
    }
}

/// Used by an `ArrayVisitor` to deserialize every element of a JS array
/// to a SATS array.
struct ArrayAccess<'a, 's, T> {
    common: DeserializerCommon<'a, 's>,
    arr: Local<'s, Array>,
    seeds: RepeatN<T>,
    index: u32,
}

impl<'de, 'a, 's, T> ArrayAccess<'a, 's, T>
where
    T: DeserializeSeed<'de> + Clone,
{
    fn new(arr: Local<'s, Array>, common: DeserializerCommon<'a, 's>, seed: T) -> Self {
        Self {
            arr,
            common,
            seeds: repeat_n(seed, arr.length() as usize),
            index: 0,
        }
    }
}

impl<'de, 's: 'de, T: DeserializeSeed<'de> + Clone> de::ArrayAccess<'de> for ArrayAccess<'_, 's, T> {
    type Element = T::Output;
    type Error = Error<'s>;

    fn next_element(&mut self) -> Result<Option<Self::Element>, Self::Error> {
        self.seeds
            .next()
            .map(|seed| {
                // Extract the array element.
                let val = self
                    .arr
                    .get_index(self.common.scope, self.index)
                    .ok_or_else(exception_already_thrown)?;

                // Deserialize the element.
                let val = seed.deserialize(Deserializer {
                    common: self.common.reborrow(),
                    input: val,
                })?;

                self.index += 1;
                Ok(val)
            })
            .transpose()
    }

    fn size_hint(&self) -> Option<usize> {
        Some(self.seeds.len())
    }
}
