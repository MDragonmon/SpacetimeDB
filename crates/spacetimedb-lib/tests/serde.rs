use serde::de::DeserializeSeed;
use spacetimedb_lib::{ElementDef, EnumDef, TupleDef, TupleValue, TypeDef};

macro_rules! de_json_snapshot {
    ($schema:expr, $json:expr) => {
        let (schema, json) = (&$schema, &$json);
        let value = de_json(schema, json).unwrap();
        let value = format!("{value:#}");
        let debug_expr = format!("de_json({})", json.trim());
        insta::assert_snapshot!(insta::internals::AutoName, value, &debug_expr);
    };
}

#[test]
fn test_json_mappings() {
    let schema = tuple(
        "args",
        [
            ("foo", TypeDef::U32),
            ("bar", TypeDef::Bytes),
            ("baz", vec(TypeDef::String)),
            (
                "quux",
                TypeDef::Enum(enumm([("Hash", TypeDef::Hash), ("Unit", TypeDef::Unit)])),
            ),
            ("and_peggy", TypeDef::F64),
        ],
    );
    let data = r#"
{
    "foo": 42,
    "bar": "404040FFFF0A48656C6C6F",
    "baz": ["heyyyyyy", "hooo"],
    "quux": { "Hash": "54a3e6d2b0959deaacf102292b1cbd6fcbb8cf237f73306e27ed82c3153878aa" },
    "and_peggy": 3.141592653589793238426
}
"#; // all of those ^^^^^^ digits are from memory
    de_json_snapshot!(schema, data);
    let data = r#"
{
    "foo": 5654,
    "bar": "010F2C",
    "baz": ["it's ðŸ¥¶Â°C"],
    "quux": { "Unit": null },
    "and_peggy": 9.8
}
"#;
    de_json_snapshot!(schema, data);
}

fn tuple<'a>(name: &str, elems: impl IntoIterator<Item = (&'a str, TypeDef)>) -> TupleDef {
    TupleDef {
        name: Some(name.into()),
        elements: elements(elems),
    }
}
fn enumm<'a>(elems: impl IntoIterator<Item = (&'a str, TypeDef)>) -> EnumDef {
    EnumDef {
        variants: elements(elems),
    }
}

fn elements<'a>(elems: impl IntoIterator<Item = (&'a str, TypeDef)>) -> Vec<ElementDef> {
    elems
        .into_iter()
        .enumerate()
        .map(|(i, (name, element_type))| ElementDef {
            tag: i.try_into().unwrap(),
            name: Some(name.into()),
            element_type,
        })
        .collect()
}

fn vec(element_type: TypeDef) -> TypeDef {
    TypeDef::Vec {
        element_type: Box::new(element_type),
    }
}

fn de_json(schema: &TupleDef, data: &str) -> serde_json::Result<TupleValue> {
    let mut de = serde_json::Deserializer::from_str(data);
    let val = schema.deserialize(&mut de)?;
    de.end()?;
    Ok(val)
}
