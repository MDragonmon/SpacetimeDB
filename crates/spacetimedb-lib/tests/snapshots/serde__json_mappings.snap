---
source: crates/spacetimedb-lib/tests/serde.rs
expression: "de_json({\n    \"foo\": 42,\n    \"bar\": \"404040FFFF0A48656C6C6F\",\n    \"baz\": [\"heyyyyyy\", \"hooo\"],\n    \"quux\": { \"Hash\": \"54a3e6d2b0959deaacf102292b1cbd6fcbb8cf237f73306e27ed82c3153878aa\" },\n    \"and_peggy\": 3.141592653589793238426\n})"
---
<tuple> {
    0: <u32> 42,
    1: <bytes> "@@@\xff\xff\nHello",
    2: <list> [
        <string> "heyyyyyy",
        <string> "hooo",
    ],
    3: <enum>.0: <hash> 54a3e6d2b0959deaacf102292b1cbd6fcbb8cf237f73306e27ed82c3153878aa,
    4: <f64> 3.141592653589793,
}
